### PR TITLE
backupccl: add protected ts to BACKUPs run in DETACHED mode

### LINKED ISSUE #51067: backup: DETACHED backups don't protect timestamp
We need to protect in this branch too -- currently it is only done below, after the job creation. We can even do it in same txn used to plan the backup when doing DETACHED.

### PR DESCRIPTION
**Pull request by adityamaru**
Fixes: #51067

Release note: None

### CODE REVIEW

File: pkg/ccl/backupccl/backup_planning.go [hunk 1]
```diff
@@ -1051,7 +1051,17 @@ func backupPlanHook(
 			CreatedBy: backupStmt.CreatedByInfo,
 		}
 
+		var sj *jobs.StartableJob
 		if backupStmt.Options.Detached {
+			// The protect timestamp logic for a DETACHED BACKUP can be run within the
+			// same txn as the BACKUP is being planned in, because we do not wait for
+			// the BACKUP job to complete.
+			sj, err = protectTimestampForBackup(ctx, p, jr, p.ExtendedEvalContext().Txn, resultsCh, spans,
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:1059]: your new helper creates the job, right?
but doesn't `StartAsyncJob` also create a job below? I'm a bit confused

Honestly I'm not sure what `StartAsyncJob` is buying us -- i might just inline it here/in restore instead, so you could do something like

```
var resRow Datums
if err := db.Txn(txn => 
  j := CreateAdoptableJob(txn) 
  protectTS(txn, j.ID)
); err != nil {
return err 
}
resultsCh <- resRow
```

File: pkg/ccl/backupccl/backup_planning.go [hunk 2]
```diff
@@ -1095,6 +1094,35 @@ func backupPlanHook(
...
+	ctx context.Context,
+	p sql.PlanHookState,
+	jr jobs.Record,
+	txn *kv.Txn,
+	resultsCh chan<- tree.Datums,
+	spans []roachpb.Span,
+	startTime, endTime hlc.Timestamp,
+	backupDetails jobspb.BackupDetails,
+) (*jobs.StartableJob, error) {
+	sj, err := p.ExecCfg().JobRegistry.CreateStartableJobWithTxn(ctx, jr, txn, resultsCh)
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:1107]: Do we want this here? don't the two flavors use different job creation methods? I suspect we want just the `len(spans)... ` onwand and to pass the job ID

**Review by adityamaru — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52914)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt and @pbardea)

___
*[pkg/ccl/backupccl/backup_planning.go, line 1059 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52914#-MExaCfKpH2PHpDIn1R--r1-1059:-MEy-W5WEYuL-BT0Gsmt:b-bp90ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/64501af6276f05496bd0578ba83c3e621ae31b38/pkg/ccl/backupccl/backup_planning.go#L1059)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

your new helper creates the job, right?
but doesn't `StartAsyncJob` also create a job below? I'm a bit confused

Honestly I'm not sure what `StartAsyncJob` is buying us -- i might just inline it here/in restore instead, so you could do something like

```
var resRow Datums
if err := db.Txn(txn => 
  j := CreateAdoptableJob(txn) 
  protectTS(txn, j.ID)
); err != nil {
return err 
}
resultsCh <- resRow
```
</blockquote></details>

Fixed and inlined the method. Used the planner txn instead of creating a new db.Txn because that was what `StartAsyncJob` did.
___
*[pkg/ccl/backupccl/backup_planning.go, line 1107 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52914#-MExaCfKpH2PHpDIn1R--r1-1107:-MExzora5un-a43MAdm5:b-tz4nk) ([raw file](https://github.com/cockroachdb/cockroach/blob/64501af6276f05496bd0578ba83c3e621ae31b38/pkg/ccl/backupccl/backup_planning.go#L1107)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

Do we want this here? don't the two flavors use different job creation methods? I suspect we want just the `len(spans)... ` onwand and to pass the job ID
</blockquote></details>

yikes sorry, this is not what I thought I did. Fixed.

File: pkg/ccl/backupccl/backup_test.go [hunk 3]
```diff
@@ -5022,80 +5022,110 @@ func TestProtectedTimestampsDuringBackup(t *testing.T) {
...
+		{
+			"backup-normal",
+			func(t *testing.T, query string, sqlDB *sqlutils.SQLRunner) {
+				sqlDB.Exec(t, query)
+			},
+		},
+		{
+			"backup-detached",
+			func(t *testing.T, query string, sqlDB *sqlutils.SQLRunner) {
+				backupWithDetachedOption := query + `WITH DETACHED`
```
Comment (pbardea) [inline pkg/ccl/backupccl/backup_test.go:5039]: nit: shouldn't there be a space before "WITH"?

**Review by pbardea — APPROVED** [review]
LGTM

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/backup_test.go (hunk 3 above)
Comment (adityamaru) [reply pkg/ccl/backupccl/backup_test.go]: done.

**Comment by adityamaru** [comment]
TFTR!

bors r+

**Pull request merged**